src/main/java/de/zalando/aruha/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -0,0 +1,366 @@
+        releasingPartitions.keySet().stream()
+                .filter(p -> !newReassigning.containsKey(p))
+                .collect(Collectors.toList()).forEach(releasingPartitions::remove);
+
+        // 3. Add releasing partitions information
+        if (!newReassigning.isEmpty()) {
+            newReassigning.keySet().forEach(this::addPartitionToReassigned);
+        }
+
+        // 4. Select which partitions must be added and add it.
+        newAssigned.values()
+                .stream()
+                .filter(p -> !offsets.containsKey(p.getKey()))
+                .forEach(this::addToStreaming);
+        // 5. Check if something can be released right now
+        reassignCommited();
+        // 6. Reconfigure kafka consumer
+        reconfigureKafkaConsumer(false);
+    }
+
+    private void addPartitionToReassigned(final Partition.PartitionKey partitionKey) {
+        final long currentTime = System.currentTimeMillis();
+        final long barrier = currentTime + getParameters().commitTimeoutMillis;
+        releasingPartitions.put(partitionKey, barrier);
+        scheduleTask(() -> barrierOnRebalanceReached(partitionKey), getParameters().commitTimeoutMillis, TimeUnit.MILLISECONDS);
+    }
+
+    private void barrierOnRebalanceReached(final Partition.PartitionKey pk) {
+        if (!isCurrent() || !releasingPartitions.containsKey(pk)) {
+            return;
+        }
+        final long currentTime = System.currentTimeMillis();
+        if (currentTime >= releasingPartitions.get(pk)) {
+            shutdownGracefully();
+        } else {
+            // Schedule again, probably something happened (ex. rebalance twice)
+            scheduleTask(() -> barrierOnRebalanceReached(pk), releasingPartitions.get(pk) - currentTime, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private void reconfigureKafkaConsumer(final boolean forceSeek) {
+        final Set<Partition.PartitionKey> currentAssignment = kafkaConsumer.assignment().stream()
+                .map(tp -> new Partition.PartitionKey(tp.topic(), String.valueOf(tp.partition())))
+                .collect(Collectors.toSet());
+        if (!currentAssignment.equals(offsets.keySet()) || forceSeek) {
+            final Map<Partition.PartitionKey, TopicPartition> kafkaKeys = offsets.keySet().stream().collect(
+                    Collectors.toMap(
+                            k -> k,
+                            k -> new TopicPartition(k.topic, Integer.valueOf(k.partition))));
+            // Ignore order
+            kafkaConsumer.assign(new ArrayList<>(kafkaKeys.values()));
+            // Check if offsets are available in kafka
+            kafkaConsumer.seekToBeginning(kafkaKeys.values().toArray(new TopicPartition[kafkaKeys.size()]));
+            kafkaKeys.forEach((key, kafka) -> offsets.get(key).ensureDataAvailable(kafkaConsumer.position(kafka)));
+            //
+            kafkaKeys.forEach((k, v) -> kafkaConsumer.seek(v, offsets.get(k).getSentOffset()));
```

Comment:
There is a bug here that will result in double sending of message. 
"offsets" are the offsets that are committed (at least in the scenario when you start stream for the partition) and when you call seek() with these offsets that will deliver you the last committed event again. 
(User committed offset 4391, and when you do seek(x, 4391) - you will then receive event with offset 4391 again from kafkaConsumer)